File: App/Resources/AlgorithmDetails/inplacelsdradixsort/inplacelsdradixsort.c
#include <stdio.h>
#include <stdlib.h>

int array[16] = {0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56};

void swap(int *a, int *b) {
  int t = *a;
  *a = *b;
  *b = t;
}

void printList(int items[], int size) {
  for (int i = 0; i < size; i++) {
    if (i == 0) {
      printf("[%d, ", items[i]);
    } else if (i != size - 1) {
      printf("%d, ", items[i]);
    } else {
      printf("%d]", items[i]);
    }
  }
}

int intPow(int base, int exponent) {
  int result = 1;
  for (int i = 0; i < exponent; i++) {
    result *= base;
  }
  return result;
}

int getDigit(int value, int power, int radix) {
  return (value / intPow(radix, power)) % radix;
}

void multiSwap(int arr[], int pos, int to) {
  if (to > pos) {
    for (int k = pos; k < to; k++) {
      swap(&arr[k], &arr[k + 1]);
    }
  } else if (to < pos) {
    for (int k = pos; k > to; k--) {
      swap(&arr[k], &arr[k - 1]);
    }
  }
}

void sort(int arr[], int n) {
  if (n == 0) {
    return;
  }
  int radix = 4;
  int maxValue = arr[0];
  for (int i = 1; i < n; i++) {
    if (arr[i] > maxValue) {
      maxValue = arr[i];
    }
  }

  int maxPower = 0;
  int probe = radix;
  while (probe <= maxValue) {
    maxPower++;
    probe *= radix;
  }

  int vregs[radix - 1]; // one register per nonzero digit value (1..radix-1)

  for (int power = 0; power <= maxPower; power++) {
    for (int i = 0; i < radix - 1; i++) {
      vregs[i] = n - 1;
    }

    int pos = 0;
    for (int step = 0; step < n; step++) {
      int digit = getDigit(arr[pos], power, radix);
      if (digit == 0) {
        pos++;
      } else {
        int to = vregs[digit - 1];
        multiSwap(arr, pos, to);
        for (int j = digit - 1; j > 0; j--) {
          vregs[j - 1]--;
        }
      }
    }
  }
}

int main(int argc, char *argv[]) {
  int size = sizeof(array) / sizeof(array[0]);
  sort(array, size);
  printList(array, size);
  return 0;
}

File: Modules/BuiltInAlgorithms/Sources/Distribution/InPlaceLSDRadixSort.swift
import AlgorithmKit
import SortEngineKit

/// An LSD radix sort that avoids `LSDRadixSort`'s full-size output array by shuffling elements
/// around in place instead — at the cost of doing quadratically more element movement per digit
/// place than that counting-sort-based approach needs. For each digit place (least significant
/// first), `vregs[d - 1]` tracks the next slot (counting in from the end of the array) where an
/// element with digit value `d` should land this pass; every register starts at `n - 1` and only
/// ever moves inward. A single left-to-right scan then does the whole pass: an element already
/// showing digit `0` is already correctly grouped ahead of every nonzero digit for this place, so
/// the scan just steps over it (`pos` advances); anything else gets walked all the way out to its
/// own register's slot via a chain of adjacent swaps (a rotation, not a single swap — everything
/// between `pos` and the target slides over by one to make room), and every *lower* nonzero
/// digit's register shifts inward by one, since the slot it used to reserve just got claimed by
/// this element's rotation passing through it.
///
/// A rotation's cost is the distance it travels, and that distance isn't bounded by a constant —
/// it depends on how far this element's target register has already drifted inward, which is
/// itself a function of how many other elements this pass has already routed past it. Repeated
/// across a whole pass, that makes one digit place's total work quadratic in `n` rather than
/// linear, and this shows up empirically (measured, not just asymptotic): a size-256 run touches
/// on the order of 150,000 swaps per full sort, regardless of whether the input started sorted or
/// adversarially shuffled — the digit-place grouping this algorithm builds has no relationship to
/// whatever order the array started in.
public struct InPlaceLSDRadixSort: SortAlgorithm {
  public let id = AlgorithmID(rawValue: "inplacelsdradixsort")
  public let metadata = AlgorithmMetadata(
    displayName: "In-Place LSD Radix Sort",
    category: .distribution,
    sizeRange: 16...256,
    growthModel: OperationGrowthModel(
      anchorSize: 141, coefficients: [237470, 3949.39, 16.6802],
      measuredSafeCeiling: nil),
    detectedGrowthModel: DetectedGrowthModel(
      family: .polynomialIntercept, coefficients: [16.6802, -754.42, 12224], rSquared: 0.999925),
    stable: true,
    timeComplexity: ComplexityBounds(
      best: "O(n^2 log n)", average: "O(n^2 log n)", worst: "O(n^2 log n)"),
    spaceComplexity: "O(b)",
    iconName: "number.square.fill"
  )

  public init() {}

  public func record(into engine: inout RecordingEngine) {
    let n = engine.count
    guard n > 1 else { return }
    let radix = 4

    func intPow(_ base: Int, _ exponent: Int) -> Int {
      var result = 1
      for _ in 0..<exponent { result *= base }
      return result
    }

    func getDigit(_ value: Int, _ power: Int) -> Int {
      (value / intPow(radix, power)) % radix
    }

    var maxValue = 0
    for i in 0..<n { maxValue = max(maxValue, engine.values[i]) }
    var maxPower = 0
    var probe = radix
    while probe <= maxValue {
      maxPower += 1
      probe *= radix
    }

    let vregHandle = engine.createAuxArray(length: radix - 1)
    var vregs = [Int](repeating: 0, count: radix - 1)

    for power in 0...maxPower {
      for i in 0..<vregs.count {
        vregs[i] = n - 1
        engine.writeAux(vregHandle, at: i, value: vregs[i])
      }

      var pos = 0
      for _ in 0..<n {
        let digit = getDigit(engine.values[pos], power)
        if digit == 0 {
          pos += 1
        } else {
          let to = vregs[digit - 1]
          if to > pos {
            for k in pos..<to { engine.swap(k, k + 1) }
          } else if to < pos {
            for k in stride(from: pos, to: to, by: -1) { engine.swap(k, k - 1) }
          }
          for j in stride(from: digit - 1, to: 0, by: -1) {
            vregs[j - 1] -= 1
            engine.writeAux(vregHandle, at: j - 1, value: vregs[j - 1])
          }
        }
      }
    }

    engine.deleteAuxArray(vregHandle)
  }
}
